pkg/cloudprovider/mesos/mesos.go

```diff
@@ -0,0 +1,238 @@
+	client *mesosClient
+	config *Config
+}
+
+func (c *MesosCloud) MasterURI() string {
+	return c.config.MesosMaster
+}
+
+func newMesosCloud(configReader io.Reader) (*MesosCloud, error) {
+	config, err := readConfig(configReader)
+	if err != nil {
+		return nil, err
+	}
+
+	log.V(1).Infof("new mesos cloud, master='%v'", config.MesosMaster)
+	if d, err := detector.New(config.MesosMaster); err != nil {
+		log.V(1).Infof("failed to create master detector: %v", err)
+		return nil, err
+	} else if cl, err := newMesosClient(d,
+		config.MesosHttpClientTimeout.Duration,
+		config.StateCacheTTL.Duration); err != nil {
+		log.V(1).Infof("failed to create mesos cloud client: %v", err)
+		return nil, err
+	} else {
+		return &MesosCloud{client: cl, config: config}, nil
+	}
+}
+
+// Instances returns a copy of the Mesos cloud Instances implementation.
+// Mesos natively provides minimal cloud-type resources. More robust cloud
+// support requires a combination of Mesos and cloud-specific knowledge.
+func (c *MesosCloud) Instances() (cloudprovider.Instances, bool) {
+	return c, true
+}
+
+// TCPLoadBalancer always returns nil, false in this implementation.
+// Mesos does not provide any type of native load balancing by default,
+// so this implementation always returns (nil, false).
+func (c *MesosCloud) TCPLoadBalancer() (cloudprovider.TCPLoadBalancer, bool) {
+	return nil, false
+}
+
+// Zones always returns nil, false in this implementation.
+// Mesos does not provide any type of native region or zone awareness,
+// so this implementation always returns (nil, false).
+func (c *MesosCloud) Zones() (cloudprovider.Zones, bool) {
+	return nil, false
+}
+
+// Clusters returns a copy of the Mesos cloud Clusters implementation.
+// Mesos does not provide support for multiple clusters.
+func (c *MesosCloud) Clusters() (cloudprovider.Clusters, bool) {
+	return c, true
+}
+
+// ListClusters lists the names of the available Mesos clusters.
+func (c *MesosCloud) ListClusters() ([]string, error) {
+	// Always returns a single cluster (this one!)
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	name, err := c.client.clusterName(ctx)
+	return []string{name}, err
+}
+
+// Master gets back the address (either DNS name or IP address) of the master node for the cluster.
+func (c *MesosCloud) Master(clusterName string) (string, error) {
+	clusters, err := c.ListClusters()
+	if err != nil {
+		return "", err
+	}
+	for _, name := range clusters {
+		if name == clusterName {
+			if c.client.master == "" {
+				return "", errors.New("The currently leading master is unknown.")
+			}
+
+			host, _, err := net.SplitHostPort(c.client.master)
+			if err != nil {
+				return "", err
+			}
+
+			return host, nil
+		}
+	}
+	return "", errors.New(fmt.Sprintf("The supplied cluster '%v' does not exist", clusterName))
+}
+
+// ipAddress returns an IP address of the specified instance.
+func (c *MesosCloud) ipAddress(name string) (net.IP, error) {
```

Comment:
Why is "c" a method receiver for this function -- it doesn't use "c" anywhere AFAICT.